module-system/module-implementations/sov-nft-module/src/call.rs

```diff
@@ -132,35 +152,46 @@ impl<C: Context> NonFungibleToken<C> {
         &self,
         token_id: u64,
         collection_name: &str,
-        collection_uri: &str,
+        token_uri: &str,
         mint_to_address: &UserAddress<C>,
         frozen: bool,
         context: &C,
-        working_set: &mut WorkingSet<C::Storage>,
+        working_set: &mut WorkingSet<C>,
     ) -> Result<CallResponse> {
-        let (collection_address, mut collection) =
-            self.get_collection_by_name(collection_name, context, working_set)?;
-        collection.exit_if_frozen()?;
-        self.exit_if_nft_exists(token_id, &collection_address, working_set)?;
-
-        let new_nft = Nft {
-            token_id,
-            collection_address: collection_address.clone(),
-            owner: mint_to_address.clone(),
-            frozen,
-            token_uri: collection_uri.to_string(),
-        };
-
-        self.nfts.set(
-            &NftIdentifier(token_id, collection_address.clone()),
-            &new_nft,
+        let (collection_address, collection_state) = Collection::get_owned_collection(
+            collection_name,
+            &self.collections,
+            context,
             working_set,
-        );
-        collection.supply += 1;
-        self.collections
-            .set(&collection_address, &collection, working_set);
-
-        Ok(CallResponse::default())
+        )?;
+        match collection_state {
+            CollectionState::Frozen(c) => bail!(
+                "Collection with name: {} , creator: {} is frozen",
+                c.get_name(),
+                c.get_creator()
+            ),
+            CollectionState::Mutable(mut collection) => {
```

Comment:
nit: It looks like this pattern of: 
- get collection
- match
- bail if frozen
... is repeated a few times. Maybe a helper function would be good?